plugins/forum/plugin_forum.cppo.ml

```diff
@@ -10,23 +10,27 @@ open Config
 
 let ns = "forum"
 
-let wrapper fn _ conf base =
-  match base with
-  | Some b when List.assoc_opt "disable_forum" conf.base_env <> Some "yes" ->
-    fn conf b ;
-    true
-  | _ ->
+let wrapper fn conf base =
+  if List.assoc_opt "disable_forum" conf.base_env <> Some "yes" then
+    fn conf base
+  else Hutil.incorrect_request conf;
+  true
+
+let w_base =
+  let none conf =
     Hutil.incorrect_request conf ;
     true
+  in
+  Gwd_lib.Request.w_base ~none
 
 let () =
   Gwd_lib.GwdPlugin.register ~ns
-    [ "FORUM", wrapper ForumDisplay.print
-    ; "FORUM_ADD", wrapper ForumDisplay.print_add
-    ; "FORUM_ADD_OK", wrapper ForumDisplay.print_add_ok
-    ; "FORUM_DEL", wrapper ForumDisplay.print_del
-    ; "FORUM_P_P", wrapper ForumDisplay.print_access_switch
-    ; "FORUM_SEARCH", wrapper ForumDisplay.print_search
-    ; "FORUM_VAL", wrapper ForumDisplay.print_valid
-    ; "FORUM_VIEW", wrapper ForumDisplay.print
+    [ "FORUM", (fun _assets -> w_base @@ wrapper ForumDisplay.print)
+    ; "FORUM_ADD", (fun _assets -> w_base @@ wrapper ForumDisplay.print_add)
+    ; "FORUM_ADD_OK", (fun _assets -> w_base @@ wrapper ForumDisplay.print_add_ok)
+    ; "FORUM_DEL", (fun _assets -> w_base @@ wrapper ForumDisplay.print_del)
+    ; "FORUM_P_P", (fun _assets -> w_base @@ wrapper ForumDisplay.print_access_switch)
+    ; "FORUM_SEARCH", (fun _assets -> w_base @@ wrapper ForumDisplay.print_search)
+    ; "FORUM_VAL", (fun _assets -> w_base @@ wrapper ForumDisplay.print_valid)
+    ; "FORUM_VIEW", (fun _assets -> w_base @@ wrapper ForumDisplay.print)
```

Comment:
Same point as before, up to now reading was never using locks. Maybe it should.

Reply:
Reading should not lock the base!!

Reply:
reading should take a read lock:
- other threads should not be able to write to the base while the read lock is on.
- but other threads can still read the base while the read lock is on.

Currently it is probably possible to start reading the base but then another thread start writing (and taking a lock) something to where we were reading and makes what we read invalid

Reply:
I disagree!
Even with a lock, what I read now and 5 minutes later may be different!!
For performance reasons, I am convinced reading should not lock the base.
This would be a major issue for a large base with many wizards such as Roglo (even if we have duplicated the base in two : one for friends - read-only and another one for wizards - read-write). Given the number of active wizards, their read actions should not lock the base.

Reply:
Yes, but the point is that if you don't take a read lock before reading, it is possible to have a read and a write happen at the same time, which may produce corrupted readings. Best case scenario it just makes the request fail, worst case you read improper data. Not the end of the world but not ideal. Read locks are not exclusive, they just prevent write locks to be acquired, so multiple reads can happen at the same time, obviously you can't write while someone is reading, which can be an issue if there is a lot of reads happening, eventually preventing any write from occuring (not that likely in practice, but it depends on the practical case).

Reply:
see:
https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock

Every database systems prevent to read and write at the same time.

Reply:
Roglo has 200+ wizards, 15 to 20 active at a given time. We already observe quite often a "locked base" situation due to multiple writes! Locking the base for reads will make the situation even worse!
I believe that request failure or incorrect data are preferable to "locked base" syndrome.
Ask the question to Pierfit's administrator and see what his reaction is.

Reply:
Not saying it should be done as is, or as you seem to suggest, but in the ideal situation we should be able to have only part of the base be locked instead ot the whole base whenever writes are happening for instance (same for reads), but it requires to ensure a lot of properties on the storage, such as fixed data size and maybe some others that don't come to my mind right now.

Reply:
We won't add a read lock!

... because we will have to change the database someday for something better.
Making a good database system is extremely hard and Geneweb should not have its own custom database system (made 30years ago, probably for the fun of making one, and not made for huge trees and 200+ editors).
It causes many problems such has the one this PR tries to fix.